### PR TITLE
[REST] add a rest endpoint for estimatesmartfee, docs, and test

### PR DESCRIPTION
**Pull request by joemphilips**
This could be useful if other clients want to use the core's fee estimation logic via REST.

### CODE REVIEW

File: test/functional/test_framework/rest.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import http.client
+
+
+#allows simple http get calls
+def http_get_call(host, port, path, response_object = 0):
```
Comment (MarcoFalke) [inline test/functional/test_framework/rest.py]: nit: Instead of moving them to a separate file, you could add convenience wrappers to the `test_node` class? Thus, it would not be required to pass in host and port all the time.

File: src/rest.cpp [hunk 2]
```diff
@@ -2,7 +2,6 @@
 // Copyright (c) 2009-2016 The Bitcoin Core developers
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
-
```
Comment (promag) [inline src/rest.cpp]: Unrelated change, please revert.

File: src/rest.cpp [hunk 3]
```diff
@@ -283,6 +283,7 @@ static bool rest_chaininfo(HTTPRequest* req, const std::string& strURIPart)
     case RF_JSON: {
         JSONRPCRequest jsonRequest;
         jsonRequest.params = UniValue(UniValue::VARR);
+
```
Comment (promag) [inline src/rest.cpp]: Unrelated change, please revert.

File: src/rest.cpp [hunk 4]
```diff
@@ -570,6 +571,55 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
...
+    std::transform(modestr.cbegin(), modestr.cend(), modestr.begin(), toupper);
+    FeeEstimateMode mode;
+    if (!FeeModeFromString(modestr, mode)){
+        return RESTERR(req, HTTP_BAD_REQUEST, "<MODE> must be one of <unset|economical|conservative>");
+    };
+
+
+    switch (rf) {
+    case RF_JSON: {
+
```
Comment (promag) [inline src/rest.cpp]: Remove empty line.

File: src/rest.cpp [hunk 5]
```diff
@@ -570,6 +571,55 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
...
+    // check estimation mode is valid
+    auto modestr = path[0];
+    std::transform(modestr.cbegin(), modestr.cend(), modestr.begin(), toupper);
+    FeeEstimateMode mode;
+    if (!FeeModeFromString(modestr, mode)){
+        return RESTERR(req, HTTP_BAD_REQUEST, "<MODE> must be one of <unset|economical|conservative>");
+    };
+
+
```
Comment (promag) [inline src/rest.cpp]: Remove empty line.

File: src/rest.cpp [hunk 6]
```diff
@@ -570,6 +571,55 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
...
+        JSONRPCRequest jsonRequest;
+        UniValue conf_target(std::stoi(path[1]));
+        UniValue mode(modestr);
+        UniValue params(UniValue::VARR);
+        params.push_back(conf_target);
+        params.push_back(mode);
+        jsonRequest.params = params;
+
+        UniValue estimatedfee;
+        estimatedfee = estimatesmartfee(jsonRequest);
```
Comment (promag) [inline src/rest.cpp]: Sounds weird to "redirect" to RPC handler (redo the mode parse etc). Maybe factor out what is needed from there?

Maybe just call `::feeEstimator.estimateSmartFee()` here?

File: doc/REST-interface.md [hunk 7]
```diff
@@ -98,6 +98,13 @@ Only supports JSON as output format.
 Returns transactions in the TX mempool.
 Only supports JSON as output format.
 
+#### Fees
+`GET /rest/fee/<MODE>/<TARGET>.json`
+
+Returns proper fee estimated by bitcoind. `<MODE>` should be one of `<unset|conservative|economical>`.
+`<TARGET>` is the desired confirmation time (in block height).
+
+
```
Comment (promag) [inline doc/REST-interface.md]: Remove 2nd empty line.

File: src/rest.cpp [hunk 8]
```diff
@@ -570,6 +571,55 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
...
+    if (!CheckWarmup(req))
+        return false;
+    std::string param;
+    const RetFormat rf = ParseDataFormat(param, strURIPart);
+
+    std::vector<std::string> path;
+    boost::split(path, param, boost::is_any_of("/"));
+    path.erase(path.begin());
+
+    if (path.size() != 2)
```
Comment (promag) [inline src/rest.cpp]: Missing `{}`.

File: src/rest.cpp [hunk 9]
```diff
@@ -570,6 +571,55 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
     }
 }
 
+// A bit of a hack - dependency on a function defined in rpc/mining.cpp
+UniValue estimatesmartfee(const JSONRPCRequest& request);
+
+static bool rest_getfee(HTTPRequest* req, const std::string& strURIPart) {
+    if (!CheckWarmup(req))
```
Comment (promag) [inline src/rest.cpp]: Missing `{}`.

File: doc/REST-interface.md [hunk 10]
```diff
@@ -98,6 +98,13 @@ Only supports JSON as output format.
 Returns transactions in the TX mempool.
 Only supports JSON as output format.
 
+#### Fees
+`GET /rest/fee/<MODE>/<TARGET>.json`
+
+Returns proper fee estimated by bitcoind. `<MODE>` should be one of `<unset|conservative|economical>`.
```
Comment (promag) [inline doc/REST-interface.md]: `Returns proper fee estimated by bitcoind`, this is `bitcoind`. 😄

**Review by promag — COMMENTED** [review]
Light concept ACK. Maybe too much for REST?

At least the the URI could be improved. This looks weird:
 - /rest/fee/economical/6.json
 - /rest/fee/unset/3.json

**Comment by joemphilips** [comment]
Thanks for reviewing.
I will update according to the review soon.
> Maybe too much for REST?

Rationale for this PR is that some light client(or other services) wants to use information about block or mempool for estimating fee. This is pretty close to the motivation of `getutxo` in terms of it places some trust on the full node. core's fee estimation logic is likely to update to a more sophisticated scheme in the near future so many clients may want to use core's logic directly.
> At least the the URI could be improved. This looks weird:

You are right. Can I use `&foo=bar` style query parameter here? I didn't because other endpoint didn't. in that case, what about
`/rest/fee.json&target=<TARGET>&mode=<MODE>` and make `mode` optional? Though I'm not sure how easy this is to implement. 

Otherwise all I can think is enabling to omit `<MODE>` . and query in the form like
`/rest/fee/5.json`

**Comment by promag** [comment]
My vote goes to `GET /rest/fee?target=<TARGET>&mode=<MODE>`, with query parameters and without extension. IMO all endpoints should return JSON, unless `Accept: ` headers requests other format.

File: src/rest.cpp [hunk 11]
```diff
@@ -570,6 +571,55 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
     }
 }
 
+// A bit of a hack - dependency on a function defined in rpc/mining.cpp
+UniValue estimatesmartfee(const JSONRPCRequest& request);
```
Comment (jonasschnelli) [inline src/rest.cpp]: I think it's not worth passing through the RPC call `estimatesmartfee`.
Better directly use `feeEstimator.estimateFee(nBlocks);`.

**Comment by jonasschnelli** [comment]
Concept ACK.
Please no query string.
The in this PR proposed URL scheme seems correct and in align with other calls: https://github.com/bitcoin/bitcoin/blob/master/doc/REST-interface.md#query-utxo-set

File: test/functional/smartfees.py [hunk 12]
```diff
@@ -117,7 +120,13 @@ def check_estimates(node, fees_seen, max_invalid, print_estimates = True):
         if e >= 0:
             valid_estimate = True
             if i >= 13:  # for n>=14 estimatesmartfee(n/2) should be at least as high as estimatefee(n)
-                assert(node.estimatesmartfee((i+1)//2)["feerate"] > float(e) - delta)
+                rpcfeerate = node.estimatesmartfee((i+1)//2)["feerate"]
+                assert(rpcfeerate > float(e) - delta)
+
+                json_string = http_get_call(url.hostname, url.port, '/rest/fee/' + 'unset/' + "{}".format((i+1)//2) + '.json')
```
Comment (jb55) [inline test/functional/smartfees.py]: nit. why not:
```python
"/rest/fee/unset/{}.json".format((i+1)//2)
```

**Comment by laanwj** [comment]
```
Please no query string.
The in this PR proposed URL scheme seems correct and in align with other calls: 
```
Yep, agree.
Concept ACK.

**Review by joemphilips — COMMENTED** [review]
updated according to review by @promag , @jb55 and @MarcoFalke . Ready for next review.

**Comment by promag** [comment]
Needs rebase.

**Comment by joemphilips** [comment]
rebased.

**Comment by joemphilips** [comment]
umm... sorry, does "Needs rebase" imply needs for squashing? or just rebasing onto master?
I only did rebasing onto master but it needs squashing before merge.
(Sorry for silly question. Forgive me, this is my first PR.) @promag

**Comment by WorkShop-Office** [comment]
I do not understand what you are talking about



Sent from my Mi phone
On Joe Miyamoto <notifications@github.com>, Feb 15, 2018 3:57 PM wrote:

umm... sorry, does "Needs rebase" imply needs for squashing? or just rebasing onto master?
I only did rebasing onto master but it needs squashing before merge.
(Sorry for silly question. Forgive me, this is my first PR.) @promag<https://github.com/promag>

**Comment by fanquake** [comment]
@joemphilips You'll need to rebase onto master, as this currently has merge conflicts in `test/functional/feature_fee_estimation.py`.

At the same time, you can also squash your changes into a single commit. See [Squashing Commits](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits). Use a descriptive commit message i.e "rest: add an endpoint for estimatesmartfee".

**Comment by joemphilips** [comment]
rebased (and squashed)

File: test/functional/test_framework/test_node.py [hunk 13]
```diff
@@ -13,6 +13,8 @@
 import re
 import subprocess
 import time
+import urllib.parse
+import http.client
```
Comment (fanquake) [inline test/functional/test_framework/test_node.py]: http.client is already imported above.

**Comment by Vasyo** [comment]
Stop send me

File: test/functional/test_framework/test_node.py (hunk 13 above)
Comment (joemphilips) [reply test/functional/test_framework/test_node.py]: Thanks, done.

**Review by jgarzik — APPROVED** [review]
concept ACK

**Comment by MarcoFalke** [comment]
Still needs rebase.

Could do something like this:

```sh
git checkout rest_fee
git fetch bitcoin
git merge bitcoin/master
git reset --soft bitcoin/master
git commit -m '[REST] add a rest endpoint for estimatesmartfee, docs, and test'
git push origin rest_fee -f

**Comment by joemphilips** [comment]
rebased

File: src/rest.cpp [hunk 14]
```diff
@@ -577,6 +578,72 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
...
+            return RESTERR(req, HTTP_BAD_REQUEST, "<MODE> must be one of <unset|economical|conservative>");
+        };
+
+        // type conversions for estimateSmartFee
+        bool conservative;
+        if (mode == FeeEstimateMode::ECONOMICAL) {
+            conservative = false;
+        } else {
+            conservative = true;
+        }
```
Comment (practicalswift) [inline src/rest.cpp]: Nit: `bool conservative = mode != FeeEstimateMode::ECONOMICAL;` instead? :-)

File: src/rest.cpp [hunk 15]
```diff
@@ -577,6 +578,72 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
...
+        if (path.size() != 2) {
+            return RESTERR(req, HTTP_BAD_REQUEST, "Path must be /rest/fee/<MODE>/<TARGET>.json");
+        }
+        // check estimation mode is valid
+        auto modestr = path[0];
+        std::transform(modestr.cbegin(), modestr.cend(), modestr.begin(), ToUpper);
+        FeeEstimateMode mode;
+        if (!FeeModeFromString(modestr, mode)){
+            return RESTERR(req, HTTP_BAD_REQUEST, "<MODE> must be one of <unset|economical|conservative>");
+        };
```
Comment (practicalswift) [inline src/rest.cpp]: Drop `;` :-)

File: src/rest.cpp [hunk 16]
```diff
@@ -577,6 +578,72 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
...
+            conservative = false;
+        } else {
+            conservative = true;
+        }
+        int64_t conf_target;
+        try {
+             conf_target = atoi64(path[1]);
+        } catch (std::invalid_argument& e) {
+            return RESTERR(req, HTTP_BAD_REQUEST, strprintf("Uneable to parse confirmation target to int"));
+        };
```
Comment (practicalswift) [inline src/rest.cpp]: Drop `;` :-)

File: src/rest.cpp [hunk 17]
```diff
@@ -577,6 +578,72 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
...
+        unsigned int max_target = ::feeEstimator.HighestTargetTracked(FeeEstimateHorizon::LONG_HALFLIFE);
+        if (conf_target < 1 || (unsigned int)conf_target > max_target) {
+            return RESTERR(req, HTTP_BAD_REQUEST, strprintf("Invalid confirmation target, must be in between %u - %u", 1, max_target));
+        };
```
Comment (practicalswift) [inline src/rest.cpp]: Drop `;` :-)

File: test/functional/interface_rest.py [hunk 18]
```diff
@@ -299,5 +299,17 @@ def run_test(self):
         json_obj = self.test_rest_request("/chaininfo")
         assert_equal(json_obj['bestblockhash'], bb_hash)
 
+        # Prepate for Fee estimation
```
Comment (practicalswift) [inline test/functional/interface_rest.py]: "Prepare" :-)

File: src/rest.cpp [hunk 19]
```diff
@@ -577,6 +578,72 @@ static bool rest_getutxos(HTTPRequest* req, const std::string& strURIPart)
...
+        if (mode == FeeEstimateMode::ECONOMICAL) {
+            conservative = false;
+        } else {
+            conservative = true;
+        }
+        int64_t conf_target;
+        try {
+             conf_target = atoi64(path[1]);
+        } catch (std::invalid_argument& e) {
+            return RESTERR(req, HTTP_BAD_REQUEST, strprintf("Uneable to parse confirmation target to int"));
```
Comment (practicalswift) [inline src/rest.cpp]: "Unable" :-)

**Review by practicalswift — CHANGES_REQUESTED** [review]

**Comment by practicalswift** [comment]
This PR does not compile when rebased on `master`. Is it still active or should it be closed? :-)

**Comment by joemphilips** [comment]
> This PR does not compile when rebased on `master`. Is it still active or should it be closed? :-)

I will close. sorry for leaving it here for so long.
